Build error on ROS 2 Buildfarm for Rolling

Rolling binary jobs have been failing and as a result, the Rolling package sync is blocked.

See: https://build.ros2.org/view/Rbin_uJ64/job/Rbin_uJ64__message_tf_frame_transformer__ubuntu_jammy_amd64__binary/64/console

Hi, sorry for the delayed response.

To be honest, I don't quite know what's going on. The rolling jobs started failing out of nowhere, definitely no changes on my side. The error log indicates compilation issues withing `rclcpp/client.hpp` itself, see below.

I have also tried to reproduce the problem, but the following compiles without any issues in a `ros:rolling` Docker container.

```
docker run --rm -it ros:rolling
# ---
mkdir -p ws/src
cd ws/src
git clone https://github.com/ika-rwth-aachen/message_tf_frame_transformer.git
cd ..
colcon build
```

---

```
08:11:34 /opt/ros/rolling/include/rclcpp/rclcpp/any_subscription_callback.hpp:416:25: error: no match for ‘operator=’ (operand types are ‘rclcpp::detail::AnySubscriptionCallbackHelper<rclcpp::SerializedMessage, std::allocator<void>, false, true>::variant_type’ {aka ‘std::variant<std::function<void(const rclcpp::SerializedMessage&)>, std::function<void(const rclcpp::SerializedMessage&, const rclcpp::MessageInfo&)>, std::function<void(std::unique_ptr<rclcpp::SerializedMessage, std::default_delete<rclcpp::SerializedMessage> >)>, std::function<void(std::unique_ptr<rclcpp::SerializedMessage, std::default_delete<rclcpp::SerializedMessage> >, const rclcpp::MessageInfo&)>, std::function<void(std::shared_ptr<const rclcpp::SerializedMessage>)>, std::function<void(std::shared_ptr<const rclcpp::SerializedMessage>, const rclcpp::MessageInfo&)>, std::function<void(const std::shared_ptr<const rclcpp::SerializedMessage>&)>, std::function<void(const std::shared_ptr<const rclcpp::SerializedMessage>&, const rclcpp::MessageInfo&)>, std::function<void(std::shared_ptr<rclcpp::SerializedMessage>)>, std::function<void(std::shared_ptr<rclcpp::SerializedMessage>, const rclcpp::MessageInfo&)> >’} and ‘rclcpp::detail::SubscriptionCallbackTypeHelper<rclcpp::SerializedMessage, std::_Bind<void (message_tf_frame_transformer::MessageTfFrameTransformer::*(message_tf_frame_transformer::MessageTfFrameTransformer*, std::_Placeholder<1>))(const std::shared_ptr<rclcpp::SerializedMessage>&)>, void>::callback_type’ {aka ‘std::function<void(const std::shared_ptr<rclcpp::SerializedMessage>&)>’})
08:11:34   416 |       callback_variant_ = static_cast<typename scbth::callback_type>(callback);
08:11:34       |       ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
08:11:34 In file included from /opt/ros/rolling/include/rclcpp/rclcpp/client.hpp:29,
08:11:34                  from /opt/ros/rolling/include/rclcpp/rclcpp/callback_group.hpp:24,
08:11:34                  from /opt/ros/rolling/include/rclcpp/rclcpp/any_executable.hpp:20,
08:11:34                  from /opt/ros/rolling/include/rclcpp/rclcpp/memory_strategy.hpp:25,
08:11:34                  from /opt/ros/rolling/include/rclcpp/rclcpp/memory_strategies.hpp:18,
08:11:34                  from /opt/ros/rolling/include/rclcpp/rclcpp/executor_options.hpp:20,
08:11:34                  from /opt/ros/rolling/include/rclcpp/rclcpp/executor.hpp:37,
08:11:34                  from /opt/ros/rolling/include/rclcpp/rclcpp/executors/multi_threaded_executor.hpp:25,
08:11:34                  from /opt/ros/rolling/include/rclcpp/rclcpp/executors.hpp:21,
08:11:34                  from /opt/ros/rolling/include/rclcpp/rclcpp/rclcpp.hpp:172,
08:11:34                  from /tmp/binarydeb/ros-rolling-message-tf-frame-transformer-1.1.0/include/message_tf_frame_transformer/MessageTfFrameTransformer.ros2.hpp:34,
08:11:34                  from /tmp/binarydeb/ros-rolling-message-tf-frame-transformer-1.1.0/src/MessageTfFrameTransformer.ros2.cpp:30:
08:11:34 /usr/include/c++/11/variant:1461:9: note: candidate: ‘template<class _Tp> std::enable_if_t<((__exactly_once<std::variant<_Types>::__accepted_type<_Tp&&, typename std::enable_if<__not_self<_Tp&&>, void>::type> > && is_constructible_v<std::variant<_Types>::__accepted_type<_Tp&&, typename std::enable_if<__not_self<_Tp&&>, void>::type>, _Tp>) && is_assignable_v<std::variant<_Types>::__accepted_type<_Tp&&, typename std::enable_if<__not_self<_Tp&&>, void>::type>&, _Tp>), std::variant<_Types>&> std::variant<_Types>::operator=(_Tp&&) [with _Tp = _Tp; _Types = {std::function<void(const rclcpp::SerializedMessage&)>, std::function<void(const rclcpp::SerializedMessage&, const rclcpp::MessageInfo&)>, std::function<void(std::unique_ptr<rclcpp::SerializedMessage, std::default_delete<rclcpp::SerializedMessage> >)>, std::function<void(std::unique_ptr<rclcpp::SerializedMessage, std::default_delete<rclcpp::SerializedMessage> >, const rclcpp::MessageInfo&)>, std::function<void(std::shared_ptr<const rclcpp::SerializedMessage>)>, std::function<void(std::shared_ptr<const rclcpp::SerializedMessage>, const rclcpp::MessageInfo&)>, std::function<void(const std::shared_ptr<const rclcpp::SerializedMessage>&)>, std::function<void(const std::shared_ptr<const rclcpp::SerializedMessage>&, const rclcpp::MessageInfo&)>, std::function<void(std::shared_ptr<rclcpp::SerializedMessage>)>, std::function<void(std::shared_ptr<rclcpp::SerializedMessage>, const rclcpp::MessageInfo&)>}]’
08:11:34  1461 |         operator=(_Tp&& __rhs)
08:11:34       |         ^~~~~~~~
08:11:34 /usr/include/c++/11/variant:1461:9: note:   template argument deduction/substitution failed:
08:11:34 /usr/include/c++/11/variant: In substitution of ‘template<class ... _Types> template<class _Tp, class> using __accepted_type = std::variant<_Types>::__to_type<__accepted_index<_Tp> > [with _Tp = std::function<void(const std::shared_ptr<rclcpp::SerializedMessage>&)>&&; <template-parameter-2-2> = void; _Types = {std::function<void(const rclcpp::SerializedMessage&)>, std::function<void(const rclcpp::SerializedMessage&, const rclcpp::MessageInfo&)>, std::function<void(std::unique_ptr<rclcpp::SerializedMessage, std::default_delete<rclcpp::SerializedMessage> >)>, std::function<void(std::unique_ptr<rclcpp::SerializedMessage, std::default_delete<rclcpp::SerializedMessage> >, const rclcpp::MessageInfo&)>, std::function<void(std::shared_ptr<const rclcpp::SerializedMessage>)>, std::function<void(std::shared_ptr<const rclcpp::SerializedMessage>, const rclcpp::MessageInfo&)>, std::function<void(const std::shared_ptr<const rclcpp::SerializedMessage>&)>, std::function<void(const std::shared_ptr<const rclcpp::SerializedMessage>&, const rclcpp::MessageInfo&)>, std::function<void(std::shared_ptr<rclcpp::SerializedMessage>)>, std::function<void(std::shared_ptr<rclcpp::SerializedMessage>, const rclcpp::MessageInfo&)>}]’:
08:11:34 /usr/include/c++/11/variant:1457:14:   required by substitution of ‘template<class _Tp> std::enable_if_t<((__exactly_once<std::variant<std::function<void(const rclcpp::SerializedMessage&)>, std::function<void(const rclcpp::SerializedMessage&, const rclcpp::MessageInfo&)>, std::function<void(std::unique_ptr<rclcpp::SerializedMessage, std::default_delete<rclcpp::SerializedMessage> >)>, std::function<void(std::unique_ptr<rclcpp::SerializedMessage, std::default_delete<rclcpp::SerializedMessage> >, const rclcpp::MessageInfo&)>, std::function<void(std::shared_ptr<const rclcpp::SerializedMessage>)>, std::function<void(std::shared_ptr<const rclcpp::SerializedMessage>, const rclcpp::MessageInfo&)>, std::function<void(const std::shared_ptr<const rclcpp::SerializedMessage>&)>, std::function<void(const std::shared_ptr<const rclcpp::SerializedMessage>&, const rclcpp::MessageInfo&)>, std::function<void(std::shared_ptr<rclcpp::SerializedMessage>)>, std::function<void(std::shared_ptr<rclcpp::SerializedMessage>, const rclcpp::MessageInfo&)> >::__accepted_type<_Tp&&, typename std::enable_if<__not_self<_Tp&&>, void>::type> > && is_constructible_v<std::variant<std::function<void(const rclcpp::SerializedMessage&)>, std::function<void(const rclcpp::SerializedMessage&, const rclcpp::MessageInfo&)>, std::function<void(std::unique_ptr<rclcpp::SerializedMessage, std::default_delete<rclcpp::SerializedMessage> >)>, std::function<void(std::unique_ptr<rclcpp::SerializedMessage, std::default_delete<rclcpp::SerializedMessage> >, const rclcpp::MessageInfo&)>, std::function<void(std::shared_ptr<const rclcpp::SerializedMessage>)>, std::function<void(std::shared_ptr<const rclcpp::SerializedMessage>, const rclcpp::MessageInfo&)>, std::function<void(const std::shared_ptr<const rclcpp::SerializedMessage>&)>, std::function<void(const std::shared_ptr<const rclcpp::SerializedMessage>&, const rclcpp::MessageInfo&)>, std::function<void(std::shared_ptr<rclcpp::SerializedMessage>)>, std::function<void(std::shared_ptr<rclcpp::SerializedMessage>, const rclcpp::MessageInfo&)> >::__accepted_type<_Tp&&, typename std::enable_if<__not_self<_Tp&&>, void>::type>, _Tp>) && is_assignable_v<std::variant<std::function<void(const rclcpp::SerializedMessage&)>, std::function<void(const rclcpp::SerializedMessage&, const rclcpp::MessageInfo&)>, std::function<void(std::unique_ptr<rclcpp::SerializedMessage, std::default_delete<rclcpp::SerializedMessage> >)>, std::function<void(std::unique_ptr<rclcpp::SerializedMessage, std::default_delete<rclcpp::SerializedMessage> >, const rclcpp::MessageInfo&)>, std::function<void(std::shared_ptr<const rclcpp::SerializedMessage>)>, std::function<void(std::shared_ptr<const rclcpp::SerializedMessage>, const rclcpp::MessageInfo&)>, std::function<void(const std::shared_ptr<const rclcpp::SerializedMessage>&)>, std::function<void(const std::shared_ptr<const rclcpp::SerializedMessage>&, const rclcpp::MessageInfo&)>, std::function<void(std::shared_ptr<rclcpp::SerializedMessage>)>, std::function<void(std::shared_ptr<rclcpp::SerializedMessage>, const rclcpp::MessageInfo&)> >::__accepted_type<_Tp&&, typename std::enable_if<__not_self<_Tp&&>, void>::type>&, _Tp>), std::variant<std::function<void(const rclcpp::SerializedMessage&)>, std::function<void(const rclcpp::SerializedMessage&, const rclcpp::MessageInfo&)>, std::function<void(std::unique_ptr<rclcpp::SerializedMessage, std::default_delete<rclcpp::SerializedMessage> >)>, std::function<void(std::unique_ptr<rclcpp::SerializedMessage, std::default_delete<rclcpp::SerializedMessage> >, const rclcpp::MessageInfo&)>, std::function<void(std::shared_ptr<const rclcpp::SerializedMessage>)>, std::function<void(std::shared_ptr<const rclcpp::SerializedMessage>, const rclcpp::MessageInfo&)>, std::function<void(const std::shared_ptr<const rclcpp::SerializedMessage>&)>, std::function<void(const std::shared_ptr<const rclcpp::SerializedMessage>&, const rclcpp::MessageInfo&)>, std::function<void(std::shared_ptr<rclcpp::SerializedMessage>)>, std::function<void(std::shared_ptr<rclcpp::SerializedMessage>, const rclcpp::MessageInfo&)> >&> std::variant<std::function<void(const rclcpp::SerializedMessage&)>, std::function<void(const rclcpp::SerializedMessage&, const rclcpp::MessageInfo&)>, std::function<void(std::unique_ptr<rclcpp::SerializedMessage, std::default_delete<rclcpp::SerializedMessage> >)>, std::function<void(std::unique_ptr<rclcpp::SerializedMessage, std::default_delete<rclcpp::SerializedMessage> >, const rclcpp::MessageInfo&)>, std::function<void(std::shared_ptr<const rclcpp::SerializedMessage>)>, std::function<void(std::shared_ptr<const rclcpp::SerializedMessage>, const rclcpp::MessageInfo&)>, std::function<void(const std::shared_ptr<const rclcpp::SerializedMessage>&)>, std::function<void(const std::shared_ptr<const rclcpp::SerializedMessage>&, const rclcpp::MessageInfo&)>, std::function<void(std::shared_ptr<rclcpp::SerializedMessage>)>, std::function<void(std::shared_ptr<rclcpp::SerializedMessage>, const rclcpp::MessageInfo&)> >::operator=<_Tp>(_Tp&&) [with _Tp = std::function<void(const std::shared_ptr<rclcpp::SerializedMessage>&)>]’
08:11:34 /opt/ros/rolling/include/rclcpp/rclcpp/any_subscription_callback.hpp:416:25:   required from ‘rclcpp::AnySubscriptionCallback<MessageT, AllocatorT> rclcpp::AnySubscriptionCallback<MessageT, AllocatorT>::set(CallbackT) [with CallbackT = std::_Bind<void (message_tf_frame_transformer::MessageTfFrameTransformer::*(message_tf_frame_transformer::MessageTfFrameTransformer*, std::_Placeholder<1>))(const std::shared_ptr<rclcpp::SerializedMessage>&)>; MessageT = rclcpp::SerializedMessage; AllocatorT = std::allocator<void>]’
08:11:34 /opt/ros/rolling/include/rclcpp/rclcpp/create_generic_subscription.hpp:70:32:   required from ‘std::shared_ptr<rclcpp::GenericSubscription> rclcpp::create_generic_subscription(rclcpp::node_interfaces::NodeTopicsInterface::SharedPtr, const string&, const string&, const rclcpp::QoS&, CallbackT&&, const rclcpp::SubscriptionOptionsWithAllocator<AllocatorT>&) [with CallbackT = std::_Bind<void (message_tf_frame_transformer::MessageTfFrameTransformer::*(message_tf_frame_transformer::MessageTfFrameTransformer*, std::_Placeholder<1>))(const std::shared_ptr<rclcpp::SerializedMessage>&)>; AllocatorT = std::allocator<void>; rclcpp::node_interfaces::NodeTopicsInterface::SharedPtr = std::shared_ptr<rclcpp::node_interfaces::NodeTopicsInterface>; std::string = std::__cxx11::basic_string<char>]’
08:11:34 /opt/ros/rolling/include/rclcpp/rclcpp/node_impl.hpp:233:45:   required from ‘std::shared_ptr<rclcpp::GenericSubscription> rclcpp::Node::create_generic_subscription(const string&, const string&, const rclcpp::QoS&, CallbackT&&, const rclcpp::SubscriptionOptionsWithAllocator<AllocatorT>&) [with CallbackT = std::_Bind<void (message_tf_frame_transformer::MessageTfFrameTransformer::*(message_tf_frame_transformer::MessageTfFrameTransformer*, std::_Placeholder<1>))(const std::shared_ptr<rclcpp::SerializedMessage>&)>; AllocatorT = std::allocator<void>; std::string = std::__cxx11::basic_string<char>]’
08:11:34 /tmp/binarydeb/ros-rolling-message-tf-frame-transformer-1.1.0/src/MessageTfFrameTransformer.ros2.cpp:125:52:   required from here
```

--- a/include/message_tf_frame_transformer/MessageTfFrameTransformer.ros2.hpp
+++ b/include/message_tf_frame_transformer/MessageTfFrameTransformer.ros2.hpp
@@ -66,7 +66,7 @@ class MessageTfFrameTransformer : public rclcpp::Node {
 
   void detectMessageType();
 
-  void transformGeneric(const std::shared_ptr<rclcpp::SerializedMessage>& serialized_msg);
+  void transformGeneric(const std::shared_ptr<const rclcpp::SerializedMessage>& serialized_msg);
 
   template <typename T>
   void transform(const T& msg);
@@ -98,7 +98,7 @@ class MessageTfFrameTransformer : public rclcpp::Node {
   rclcpp::TimerBase::SharedPtr detect_message_type_timer_;
 
   rclcpp::GenericSubscription::SharedPtr subscriber_;
-  
+
   rclcpp::PublisherBase::SharedPtr publisher_;
 
   std::string msg_type_;

--- a/src/MessageTfFrameTransformer.ros2.cpp
+++ b/src/MessageTfFrameTransformer.ros2.cpp
@@ -138,7 +138,7 @@ void MessageTfFrameTransformer::detectMessageType() {
 }
 
 
-void MessageTfFrameTransformer::transformGeneric(const std::shared_ptr<rclcpp::SerializedMessage>& serialized_msg) {
+void MessageTfFrameTransformer::transformGeneric(const std::shared_ptr<const rclcpp::SerializedMessage>& serialized_msg) {
 
   if (false) {}
 #define MESSAGE_TYPE(TYPE, NAME)                                               \

